rust_src/src/dispnew.rs

```diff
@@ -143,13 +143,34 @@ pub fn internal_show_cursor_p(window: LispWindowOrSelected) -> bool {
     !win.cursor_off_p()
 }
 
+/// Return whether input is coming from the keyboard.
+// Corresponds to the INTERACTIVE macro in commands.h.
+pub fn is_interactive() -> bool {
+    unsafe { globals.Vexecuting_kbd_macro.is_nil() && !noninteractive }
+}
+
+#[no_mangle]
+pub extern "C" fn ding_internal() {
+    unsafe {
+        if noninteractive {
+            putchar_unlocked(7);
```

Comment:
Beware the Emacs source number formats. The `7` here was `07` which in C is octal. https://www.geeksforgeeks.org/octal-numbers-c/

In this case `07` is decimal 7. So it works out.